### PR TITLE
Refactor and clean up URLs and endpoints

### PR DESCRIPTION
**Pull request by dhruvkb**
## Dependencies
Please merge #160 before this.

## Description
This is a refactoring PR that cleans up, reorganises and in general improves the code quality of the endpoints. These refactors are made possible thanks to the confidence of the more comprehensive integration tests.

## Technical details
The PR makes the following improvements:
- Cleans up the URL config
  - Breaks it into more digestible chunks
  - Organises related endpoints together
  - Deprecates old endpoints, with backwards-compatible permanent redirects
- Disables the caching for Swagger documentation in dev (no need to repeatedly clean cache to see changes now!)
- Adheres to Django conventions about `settings`, `path`/`re_path` and `include`

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by zackkrida — APPROVED** [review]
almost removed as much code as was added 😄 this looks good to me! appreciate the new test for deprecated endpoints.

**Comment by dhruvkb** [comment]
@zackkrida by cleaning up I meant breaking up and reorganising big files into smaller modules.

**Comment by zackkrida** [comment]
Yes, I just noticed the diff—this is absolutely much cleaner regardless of LOC

File: openverse-api/catalog/api/utils/status_code_view.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from django.http import JsonResponse
+from django.utils.decorators import method_decorator
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic import View
+
+
+def get_status_code_view(data, status_code=200):
+    """
+    Get a class-based that returns the given data and status code on all HTTP
```
Comment (obulat) [inline openverse-api/catalog/api/utils/status_code_view.py:9]:
```suggestion
    Get a class-based view that returns the given data and status code on all HTTP
```

File: openverse-api/catalog/urls/__init__.py [hunk 2]
```diff
@@ -191,147 +186,88 @@
...
+        RedirectView.as_view(pattern_name='image-thumb', permanent=True),
+        name='thumbs'
     ),
-    re_path('images', SearchImages.as_view(), name='images'),
-    path(  # Deprecated
-        'sources',
-        RedirectView.as_view(pattern_name='image-stats', permanent=True),
-        name='about-image'
+
+    # Discontinued
```
Comment (obulat) [inline openverse-api/catalog/urls/__init__.py:232]: Is there a difference between a `deprecated` and a `discontinued` url?

**Review by obulat — APPROVED** [review]
Great changes. Nice to see code cleaned up :)

File: openverse-api/catalog/urls/__init__.py (hunk 2 above)
Comment (dhruvkb) [reply openverse-api/catalog/urls/__init__.py:232]: Deprecated URLs will be redirected to new paths, discontinued ones will return a 4xx error message.

File: openverse-api/catalog/api/utils/status_code_view.py (hunk 1 above)
Comment (dhruvkb) [reply openverse-api/catalog/api/utils/status_code_view.py]: Fixed in 80fc446, couldn't accept suggestion because of 80 character limit.

**Pull request merged**